File: node/Assembler/src/engine/preprocess/enginePreProcess.js
import { Logger } from '../../../../Arshu/src/common/Logger.js';

/**
 * PreProcess template engine implementation that delegates to ILoaderPreProcess
 * All parsing and merging logic is in the loader, engine only coordinates
 */
export class EnginePreProcess {
    constructor(appViewPrefix = '') {
        this.appViewPrefix = appViewPrefix;
    }

    /**
     * Merges templates using loader instance via ILoaderPreProcess interface
     * This method only coordinates - all logic is delegated to the loader
     * @param {string} appSite The application site name
     * @param {string} appFile The application file name
     * @param {string|null} appView The application view name (optional)
     * @param {ILoaderPreProcess} loader The loader instance that implements ILoaderPreProcess
     * @param {boolean} enableJsonProcessing Whether to enable JSON data processing
     * @returns {string} HTML with placeholders replaced
     */
    mergeTemplates(appSite, appFile, appView, loader, enableJsonProcessing = true) {
        Logger.debug(`MergeTemplates called: appSite=${appSite}, appFile=${appFile}, appView=${appView || 'null'}, enableJson=${enableJsonProcessing}`, 'EnginePreProcess');

        if (!loader) {
            Logger.warn('No loader provided', 'EnginePreProcess');
            return '';
        }

        // Get main template from loader
        const mainPreprocessed = loader.getTemplateHtml(appSite, appFile, appView, this.appViewPrefix);
        if (!mainPreprocessed) {
            Logger.warn(`Main template not found for appSite=${appSite}, appFile=${appFile}`, 'EnginePreProcess');
            return '';
        }

        Logger.debug(`Main template found, original size: ${mainPreprocessed.originalContent.length}`, 'EnginePreProcess');

        // Start with original content
        let contentHtml = mainPreprocessed.originalContent;

        // Delegate all replacement logic to the loader
        contentHtml = loader.applyAllReplacementMappings(
            contentHtml,
            appSite,
            mainPreprocessed,
            appView,
            this.appViewPrefix,
            enableJsonProcessing
        );

        Logger.debug(`MergeTemplates complete: output size=${contentHtml.length}`, 'EnginePreProcess');

        return contentHtml;
    }
}

File: rust/Assembler/src/engine/preprocess/engine_preprocess.rs
use crate::interface::ILoaderPreProcess;
use arshu::common::Logger;

/// PreProcess template engine implementation that only does merging using preprocessed data structures
/// All parsing is done by Loader, this engine only handles merging
/// Matches C# EnginePreProcess structure - uses ILoaderPreProcess for template access
pub struct EnginePreProcess {
    app_view_prefix: String,
}

impl EnginePreProcess {
    pub fn new(prefix: String) -> Self {
        Self {
            app_view_prefix: prefix,
        }
    }

    pub fn app_view_prefix(&self) -> &str {
        &self.app_view_prefix
    }

    pub fn set_app_view_prefix(&mut self, prefix: String) {
        self.app_view_prefix = prefix;
    }

    /// Merges templates using preprocessed data structures
    /// This method only does merging using preprocessed data structures - no loading or parsing
    pub fn merge_templates(
        &self,
        app_site: &str,
        app_file: &str,
        app_view: Option<&str>,
        loader: &dyn ILoaderPreProcess,
        enable_json_processing: bool,
    ) -> String {
        let app_view_str = app_view.unwrap_or("null");
        Logger::debug(
            &format!(
                "MergeTemplates called: appSite={}, appFile={}, appView={}, enableJson={}",
                app_site, app_file, app_view_str, enable_json_processing
            ),
            Some("EnginePreProcess"),
        );

        // Get main template using ILoaderPreProcess (includes AppView fallback and SearchAppSites logic)
        let main_preprocessed = match loader.get_template_html(
            app_site,
            app_file,
            app_view,
            Some(&self.app_view_prefix),
        ) {
            Some(tmpl) => tmpl,
            None => {
                Logger::warn(
                    &format!(
                        "Main template not found for appSite={}, appFile={}",
                        app_site, app_file
                    ),
                    Some("EnginePreProcess"),
                );
                return String::new();
            }
        };

        Logger::debug(
            &format!(
                "Main template found, original size: {}",
                main_preprocessed.original_content.len()
            ),
            Some("EnginePreProcess"),
        );

        // Start with original content
        let mut content_html = main_preprocessed.original_content.clone();

        // Merge JSON into main template first using loader's centralized method (for consistency with other engines)
        if enable_json_processing {
            content_html = loader.merge_html_with_json(&content_html, app_site, app_file);
            Logger::debug(
                &format!("After main template JSON merge: {} chars", content_html.len()),
                Some("EnginePreProcess"),
            );
        }

        // Apply ALL replacement mappings from ALL templates using loader's method
        // The loader handles multi-pass logic internally
        let result = loader.apply_all_replacement_mappings(
            &content_html,
            app_site,
            Some(&main_preprocessed),
            app_view,
            Some(&self.app_view_prefix),
            enable_json_processing,
        );

        Logger::debug(
            &format!("MergeTemplates complete: output size={}", result.len()),
            Some("EnginePreProcess"),
        );
        result
    }
}
